main.py

```diff
@@ -144,65 +145,79 @@ async def transform_stream_for_client(
                         continue
 
                     url_set.add(doc.metadata["source"])
-                    yield doc.metadata["title"] + ":" + doc.metadata["source"] + "\n"
-
-                if len(url_set) > 0:
-                    yield "SOURCES:----------------------------"
+                    all_sources.append(
+                        doc.metadata["title"] + ":" + doc.metadata["source"]
+                    )
+                if all_sources:
+                    src = {"sources": "\n".join(all_sources)}
+                    yield f"{json.dumps(src)}\n"
 
             elif op["path"] == "/streamed_output/-":
                 # Send stream output
-                yield op["value"]
+                yield f'{json.dumps({"tok": op["value"]})}\n'
+
+            elif not op["path"] and op["op"] == "replace":
+for                # Send final output
+                yield f'{json.dumps({"run_id": str(op["value"]["id"])})}\n'
```

Comment:
@nfcampos is this the best way to get the run ID if we are immediately returning a streaming object?

Reply:
Yep this is fine, but it would be enough for the condition to be `not op['path']`